pkg/controller/clusterinstallation/controller.go

```diff
@@ -0,0 +1,264 @@
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	err = mattermost.SetDefaults()
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if mattermost.Spec.DatabaseType.Type == "mysql" {
+		reqLogger.Info("Reconciling ClusterInstallation MySQL service account")
+		err = r.checkMySQLServiceAccount(mattermost, reqLogger)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+
+		reqLogger.Info("Reconciling ClusterInstallation MySQL role binding")
+		err = r.checkMySQLRoleBinding(mattermost, reqLogger)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+
+		reqLogger.Info("Reconciling ClusterInstallation MySQL")
+		err = r.checkMySQLDeployment(mattermost, reqLogger)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+	} else {
+		reqLogger.Info("Reconciling ClusterInstallation Postgres")
+		err = r.checkDBPostgresDeployment(mattermost, reqLogger)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+	}
+
+	// TODO
+	err = r.checkMinioDeployment(mattermost, reqLogger)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	err = r.checkMattermost(mattermost, reqLogger)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Already exists - don't requeue
+	return reconcile.Result{}, nil
+}
+
+// Object combines the interfaces that all Kubernetes objects must implement.
+type Object interface {
+	runtime.Object
+	v1.Object
+}
+
+// createResource creates the provided resource and sets the owner
+func (r *ReconcileClusterInstallation) createResource(owner v1.Object, resource Object, reqLogger logr.Logger) error {
+	if err := r.client.Create(context.TODO(), resource); err != nil {
+		reqLogger.Info("Error creating resource:", err.Error())
+		return err
+	}
+	if err := controllerutil.SetControllerReference(owner, resource, r.scheme); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r *ReconcileClusterInstallation) createServiceAccountIfNotExists(owner v1.Object, serviceAccount *corev1.ServiceAccount, reqLogger logr.Logger) error {
+	foundServiceAccount := &corev1.ServiceAccount{}
+
+	err := r.client.Get(context.TODO(), types.NamespacedName{Name: serviceAccount.Name, Namespace: serviceAccount.Namespace}, foundServiceAccount)
+	if err != nil && errors.IsNotFound(err) {
+		reqLogger.Info("Creating service account", serviceAccount.Name)
+		return r.createResource(owner, serviceAccount, reqLogger)
+	} else if err != nil {
+		reqLogger.Error(err, "Failed to check if service account exists")
+		return err
+	}
+
+	// TODO compare found service account versus expected
+
+	reqLogger.Info("Service account reconcicled")
```

Comment:
nit: maybe unify the comment above and this log into just:
```
    reqLogger.Debug("TOOD: reconcile existing service account")
```

Or maybe just remove the false log?

Reply:
Oh actually, I think we should keep these around as is even if we don't update existing resources since creating a resource when it should have existed is still successfully "reconciling"

Reply:
(Fixed the spelling mistake though)

Reply:
But if the resources doesn't exist, we create and return on line 184, so this log doesn't get emitted, right?

Reply:
Hmm right. Ok I'll remove it 👍